accepted/00062-hub-xmlrpc-api.md

```diff
@@ -33,7 +33,9 @@ Content management APIs will work as of today. Inter-Server Synchronization (ISS
 
 ### New service and endpoint
 A new XMLRPC API endpoint will be created, implemented by a new service called the "XMLRPC Gateway API" (simply called "Gateway" from now on in this document). Technology-wise:
```

Comment:
> Similarly, we don't need to _change_ the API signatures as per documentation and use multitude of auth keys from _each_ SUMA server or an avalanche of parameters in APIs, assuming there might be 500 SUMAs connected. We should have only one. And in case we would need to use a subset of SUMA servers, we could have something lile this:
> 
> ```python
> proxy.filter("ignore-me.acme.com")
> 
> ...
> 
> proxy.filter("ignore-me-too.acme.com", "and-me.acme.com")
> 
> ...
> 
> proxy.clearFilter()
> ```
> 
> All works as above, except `ignore-me.acme.com` would be... ignored completely (not even popping out in the return structure).
> 
> Does it make sense to you?

That looks to me like a different use case.

Case 1: hub administrator creates some keys for user A so that user A can act on servers 1, 3 and 5. Rationale for the apikeys: the end user should not have access to all of the servers. This is much like GitHub keys: key user and key creator are different entities, key creator does not trust key user to have full access to everything.

Case 2 (what you describe here) : hub administrator wants to perform some action on a subset of servers and limits himself to that subset. Key user may have full access but decides to drop (filter out) some of his privileges.

Reply:
Meta-note: aou are making a lot of different points here, not all related to this position in the text and some not related to this PR at all. Please divide better next time.

> essentially it is a multicast microservice

Yes.

> I would suggest that the caller method simply always returns a map string/object (host/result) 

The advantage of using ordered arrays as input parameters and ordered arrays as outputs, which is proposed here, is composability gets a bit easier. Often an API call will need input coming from output of other API calls, like this:

```python
def ids_from(struct_array):
  [e.id for e in struct_array]

systems_per_server = hub.system.listActiveSystems(key)
errata_per_server = hub.errata.findByCve(key, [cve] * n)
hub.system.scheduleApplyErrata(key, ids_from(systems_per_server), ids_from(errata_ids))
```

Using Server-indexed maps will make it more difficult to pass data out of a function and into another.

> proxy = xmlrpc.client.ServerProxy("https://hub.acme.com:12345")
> auth_key = proxy.login(USER, PASSWORD)
> result = proxy.call("system.listSystems", auth_key)

Point of this RFC is to make it easier for script developers to script the Hub, so I think the proposed syntax above is much clearer than a separate `proxy` object. Also not changing the text in this respect for the time being.

> That way anything can be called as-is on all servers simultaneously.

That is not different from what it is being proposed:

https://github.com/uyuni-project/uyuni-rfc/pull/24/files#diff-299b7fe4661063d97fab6412711ed861R36

> Exception here would be only login part, where it obviously wouldn't return all tokens to all servers! But instead it would return only one token from the Hub, and so then Hub would map that token to all the connected server tokens internally and pass per a target (keeping all that story about auto-reconnect/token-caching etc etc)

That does not sound different from what it is being proposed:

https://github.com/uyuni-project/uyuni-rfc/pull/24/files#diff-299b7fe4661063d97fab6412711ed861L54-L67


> proxy.filter("ignore-me.acme.com")
> proxy.filter("ignore-me-too.acme.com", "and-me.acme.com")
> proxy.clearFilter()
> All works as above, except ignore-me.acme.com would be... ignored completely (not even popping out in the return structure).
> Does it make sense to you?

Problem with this approach is that any call will get the same parameter values to all Servers, which might happen but it's not the most frequent case. In general, one wants to pass different parameter values to different Servers. Consider the example above: one wants to schedule an errata update on all clients - but errata ids are not guaranteed to be the same across Servers. One needs to use the right errata ids on the right Servers.


So far I do not see reasons to change the current text, I am resolving this conversation.

Reply:
> Please divide better next time.

Yes, I just reused the current PR to see if I would need to go forward. Guilty as charged.

> errata ids are not guaranteed to be the same across Servers

Right. So then I see my proposal as an optional add-on for convenience in bulk calls, beside the main text. Let's keep your text for now as is.